### ISSUE TITLE
caveman estep 1.12.0 

### ISSUE THREAD

**Issue by OCS-Consult** [cancerit/CaVEMan#77]
This error has been reported before 

[ERROR] (src/output.c: output_vcf_header:302 errno: No such file or directory) Error (0) writing contigs.
[ERROR] (src/estep.c: estep_main:517 errno: None) Error writing header to muts file.

The thread was closed suspecting a zlib error (fixed @ 1.2.8)
However we have the same error and this is on Centos 7 with zlib 1.2.7 
$ cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core)
and currently supports zlib:-
[root@server213-171-196-65 /]# rpm -qa zlib*
zlib-1.2.7-17.el7.x86_64
zlib-1.2.7-17.el7.i686
zlib-devel-1.2.7-17.el7.x86_64

As far a  I understand there isn't a version of zlib 1.2.8 available for Centos 1.2.7.

Any advice will be appreciated.

**Comment by drjsanger** [comment]
zlib >= 1.2.3.5 should be sufficient given the functionality used. 
Are you getting the correct zlib version linked used when you run 
`ldd caveman`?

**Comment by OCS-Consult** [comment]
Hi David

I got my C coder to fix the issue in zlib I believe. There is a limit to the chunck size the utility can use. We have revised the code to write out in smaller chunks and that works ok so far. I am just waiting on the users to qualify the VCF files we have produced.

Will update the forum with the full fix in due course.


Best Rgds


Colin

**Comment by ChristopheLegendre** [comment]
Hi David,

I have the same error for the caveman's step "estep". The error message is: 
` caveman estep -i 1 ` 
`[ERROR] (src/output.c: output_vcf_header:302 errno: No such file or directory) Error (0) writing contigs. 
[ERROR] (src/estep.c: estep_main:518 errno: None) Error writing header to muts file. ` 


My system is CentOS 7.2 with default gcc coming with that centOS version. The default libz is 1.2.7. 

Despite version 1.2.7> 1.2.3.5 (the minimum version), caveman raised the error with libz version 1.2.7.  
So I installed the latest version of zlib and it looks like it has been linked correctly to caveman (see below) after  I ran the command " ./setup .  " to install caveman with local htslib version provided by your script. 

`
ldd /tools/CaVEMan/bin/caveman `
`
        linux-vdso.so.1 =>  (0x00007ffdce7d4000) 
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fcfc979a000) 
        libz.so.1 => /tools/libz/1.2.11/lib/libz.so.1 (0x00007fcfc957e000) 
        libm.so.6 => /lib64/libm.so.6 (0x00007fcfc927c000) 
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fcfc9078000) 
        libc.so.6 => /lib64/libc.so.6 (0x00007fcfc8cb4000) 
        /lib64/ld-linux-x86-64.so.2 (0x00007fcfc99de000) 
` 

I do not know C programming, but it seems that functions related to libz are not recognized, such as gzprintf, gzopen

**Questions:**
- Is there any other library that should be linked to caveman?
- Should I install the version 1.2.3.5 for libz and link caveman to it?
- Would it be possible to get static binaries for caveman?

Thanks in advance for your help,
Best,
Chris

**Comment by ChristopheLegendre** [comment]
@drjsanger 
Sure.
The reference Genome is hs37d5 (GRCh37).
The contigs are **not** prefixed by "chr".
Contigs list attached.
[hs37d5.fa.chrom.contigs.txt](https://github.com/cancerit/CaVEMan/files/2349702/hs37d5.fa.chrom.contigs.txt)

**Comment by drjsanger** [comment]
Thank you @ChristopheLegendre .  

A quick update:

I have tested with a clean install of Centos 7

```
$ cat /etc/redhat-release
CentOS Linux release 7.4.1708 (Core)
```

Using yum installed libraries only:

```
$ ldd install/bin/caveman
	linux-vdso.so.1 =>  (0x00007ffc61ed3000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f58f7fe5000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f58f7dcf000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f58f7acc000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f58f78c8000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f58f74fb000)
	/lib64/ld-linux-x86-64.so.2 (0x000055c64e6a5000)
```

Which leads to CaVEMan using zlib version 1.2.7
```
$ ls -l /lib64/libz.so.1
lrwxrwxrwx. 1 root root 13 Mar  2  2018 /lib64/libz.so.1 -> libz.so.1.2.7
```

This is failing to reproduce the error when using GRCh37d5 (including all contigs to ensure largest length header possible).

It looks from your `ldd` output you have libz 1.2.11 installed and linked. 

I'll try that next. 

Dave

**Comment by drjsanger** [comment]
I have now successfully reproduced using a larger number of contigs.

Centos 7.4
zlib 1.2.11

 I'll investigate how modifying the buffer size for zlib after file opening can correct this.

**Comment by ChristopheLegendre** [comment]
Hi David,

Thanks for investigating.
For information, the extra contigs we added to our reference are very small contigs such the ERCC-xxxx, EBV and some ribosomal ones.

**Comment by drjsanger** [comment]
I've added a dynamic calculation of contig header section size to ensure the gzbuffer is now set correctly.

[13.3.3](https://github.com/cancerit/CaVEMan/releases/tag/1.13.3) contains this fix. @OCS-Consult 
apologies this took so long to get fixed!

**Comment by ChristopheLegendre** [comment]
@drjsanger 
Hi David,

No problems. I am glad I could give some feedback. 
Thanks for updating caveman.  
I have tested it and got errors; I opened a new issue #82 
Before running my new`caveman 1.13.3` executable, I checked and it is well linked to libz 1.2.11.

Best,
Chris

**Comment by OCS-Consult** [comment]
No problem David – you probably hit the issues we fixed with zlib…☺

I’ll let my chaps know.

Rgds
Colin
